storage/innobase/include/trx0purge.h

```diff
@@ -244,16 +192,25 @@ class purge_sys_t
 					record */
 	uint16_t	hdr_offset;	/*!< Header byte offset on the page */
 
-
-	TrxUndoRsegsIterator
-			rseg_iter;	/*!< Iterator to get the next rseg
-					to process */
-public:
 	purge_pq_t	purge_queue;	/*!< Binary min-heap, ordered on
 					TrxUndoRsegs::trx_no. It is protected
 					by the pq_mutex */
 	mysql_mutex_t	pq_mutex;	/*!< Mutex protecting purge_queue */
 
+public:
+  /* Push to purge queue without acquiring pq_mutex.
+  @param rseg rseg to push */
+  void push_to_purge_queue_no_lock(trx_rseg_t &rseg)
+  {
+    purge_queue.push(&rseg);
+  }
```

Comment:
I think that it would be better to just have one accessor and invoke `mysql_mutex_assert_owner(&pq_mutex)` inside it.

Reply:
I don't fully understand your idea. We need to acquire purge queue mutex outside of `purge_sys_t` in `trx_t::write_serialisation_history()`. That's why we need to separate the accessor and purge queue locking. If you propose just to add `mysql_mutex_assert_owner(&pq_mutex)` in `push_to_purge_queue_no_lock()`, then I am ok with this, but we need to manually lock/unlock the mutex in `trx_rseg_mem_restore()`, what, of course will not impact performance too much, as `trx_rseg_mem_restore()` is invoked only on server start. We could even add template argument in `push_to_purge_queue_no_lock()` function.

Reply:
Yes, that exactly is my idea. I would rename `push_to_purge_queue_no_lock()` to just `enqueue()` and omit the variant that acquires and releases the mutex. To reduce the mutex traffic on startup, we could acquire `purge_sys.pq_mutex` at the start of `trx_rseg_array_init()` and release it at the end. That should be fine from the deadlock prevention point of view, because executing that function is a prerequisite for starting the purge subsystem or any transactions. In other words, `trx_rseg_array_init()` can’t be blocked by any thread that would hold a buffer page latch while waiting for `purge_sys.pq_mutex`.

Reply:
Oh, by the way, since inserting into a priority queue is a rather complex operation, it could make sense to define the function non-`inline`.